Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -0,0 +1,150 @@
+
+namespace WebCore {
+
+VideoMediaSampleRenderer::VideoMediaSampleRenderer(AVSampleBufferDisplayLayer* layer)
+    : m_displayLayer(layer)
+{
+}
+
+VideoMediaSampleRenderer::~VideoMediaSampleRenderer()
+{
+    [m_displayLayer flush];
+    [m_displayLayer stopRequestingMediaData];
+    if (m_decompressionSession)
+        m_decompressionSession->invalidate();
+}
+
+bool VideoMediaSampleRenderer::isReadyForMoreMediaData() const
+{
+    return (!m_decompressionSession || m_decompressionSession->isReadyForMoreMediaData()) && [m_displayLayer isReadyForMoreMediaData];
+}
+
+void VideoMediaSampleRenderer::stopRequestingMediaData()
+{
+    [m_displayLayer stopRequestingMediaData];
+}
+
+void VideoMediaSampleRenderer::enqueueSample(CMSampleBufferRef sample, bool displaying)
+{
+    if (!m_decompressionSession && !m_currentCodec) {
+        // Only use a decompression session for vp8 or vp9 when software decded.
+        CMVideoFormatDescriptionRef videoFormatDescription = PAL::CMSampleBufferGetFormatDescription(sample);
+        auto fourCC = PAL::CMFormatDescriptionGetMediaSubType(videoFormatDescription);
+        if (fourCC == 'vp08' || (fourCC == 'vp09' && !(canLoad_VideoToolbox_VTIsHardwareDecodeSupported() && VTIsHardwareDecodeSupported(kCMVideoCodecType_VP9))))
+            initializeDecompressionSession();
+        m_currentCodec = fourCC;
+    }
+    if (!m_decompressionSession) {
+        [m_displayLayer enqueueSampleBuffer:sample];
+        return;
+    }
+    m_decompressionSession->enqueueSample(sample, displaying);
+    if (displaying || m_decodePending)
+        return;
+    m_decodePending = true;
+    m_decompressionSession->requestMediaDataWhenReady([weakThis = ThreadSafeWeakPtr { *this }, this] {
+        if (RefPtr protectedThis = weakThis.get()) {
+            m_decodePending = false;
+            if (m_readyForMoreSampleFunction)
+                m_readyForMoreSampleFunction();
+            m_decompressionSession->stopRequestingMediaData();
+        }
+    });
+}
+
+void VideoMediaSampleRenderer::reset()
+{
+    m_currentCodec.reset();
+    if (!m_decompressionSession)
+        return;
+    m_decompressionSession->invalidate();
+    initializeDecompressionSession();
+}
+
+void VideoMediaSampleRenderer::initializeDecompressionSession()
+{
+    m_decompressionSession = WebCoreDecompressionSession::createOpenGL();
+    m_decompressionSession->setTimebase([m_displayLayer timebase]);
+    m_decompressionSession->decodedFrameWhenAvailable([weakThis = ThreadSafeWeakPtr { *this }](RetainPtr<CMSampleBufferRef>&& sample) {
+        if (RefPtr protectedThis = weakThis.get())
+            [protectedThis->m_displayLayer enqueueSampleBuffer:sample.get()];
+    });
+    resetReadyForMoreSample();
+}
+
+void VideoMediaSampleRenderer::flush()
+{
+    [m_displayLayer flush];
+    if (m_decompressionSession)
+        m_decompressionSession->flush();
+}
+
+void VideoMediaSampleRenderer::readyForMoreSample(Function<void()>&& function)
```

Comment:
It isn't obvious what this does from the name. Maybe `whenReadyForMoreSamples`, or `requestMoreSamplesWhenReady`?

Reply:
Done